chromium_src/components/payments/content/payment_request.cc

```diff
@@ -0,0 +1,111 @@
+#include "brave/components/payments/content/buildflags/buildflags.h"
+#include "content/public/browser/browser_context.h"
+#include "third_party/blink/public/mojom/payments/payment_request.mojom.h"
+
+using payments::mojom::PaymentErrorReason;
+
+#if BUILDFLAG(ENABLE_PAY_WITH_BAT)
+namespace payments {
+
+void PaymentRequest::GetPublisherDetailsCallback(
+    mojo::PendingRemote<mojom::PaymentRequestClient> client,
+    std::vector<mojom::PaymentMethodDataPtr> method_data,
+    mojom::PaymentDetailsPtr details,
+    mojom::PaymentOptionsPtr options,
+    const ledger::type::Result result,
+    ledger::type::PublisherInfoPtr info) {
+  if (!info || info->status == ledger::type::PublisherStatus::NOT_VERIFIED) {
+    log_.Error(brave_rewards::errors::kInvalidPublisher);
+    TerminateConnection();
+    return;
+  }
+
+  Init_ChromiumImpl(std::move(client), std::move(method_data),
+                    std::move(details), std::move(options));
+}
+
+void PaymentRequest::TerminateConnectionWithMessage(PaymentErrorReason reason,
+                                                    std::string err) {
+  client_->OnError(reason, err);
+  TerminateConnection();
+}
+
+void PaymentRequest::Init(
+    mojo::PendingRemote<mojom::PaymentRequestClient> client,
+    std::vector<mojom::PaymentMethodDataPtr> method_data,
+    mojom::PaymentDetailsPtr details,
+    mojom::PaymentOptionsPtr options) {
+  bool result = false;
+  content::WebContents* contents = web_contents();
+  content::BrowserContext* context = contents->GetBrowserContext();
+
+  for (size_t i = 0; i < method_data.size(); i++) {
+    if (method_data[i]->supported_method == brave_rewards::kBatPaymentMethod) {
+      result = true;
+      break;
+    }
+  }
+
+  if (!result) {
+    Init_ChromiumImpl(std::move(client), std::move(method_data),
+                      std::move(details), std::move(options));
+    return;
+  }
+
+  // For BAT payment method, every item should have an associated SKU token
+  if (details->display_items.has_value()) {
+    for (const mojom::PaymentItemPtr& display_item : *details->display_items) {
+      if (!(display_item->sku.has_value())) {
+        log_.Error(brave_rewards::errors::kInvalidData);
+        TerminateConnection();
+        return;
+      }
+    }
+  } else {
+    log_.Error(brave_rewards::errors::kInvalidData);
+    TerminateConnection();
+    return;
+  }
+
+  // BAT payment method only works in regular mode.
+  if (context->IsOffTheRecord()) {
+    log_.Error(brave_rewards::errors::kBraveRewardsNotEnabled);
+    TerminateConnection();
+    return;
+  }
+
+  // BAT payment method only works for verified publishers
+  auto* service =
+      brave_rewards::RewardsServiceFactory::FromBrowserContext(context);
+  if (service->IsRewardsEnabled()) {
+    service->GetPublisherInfo(
+        contents->GetLastCommittedURL().GetOrigin().host(),
+        base::BindOnce(&PaymentRequest::GetPublisherDetailsCallback,
```

Comment:
there is a race condition here and this code really shouldn't be here anyway. I think it makes more sense somewhere around here https://github.com/brave/brave-core/pull/7557/files#diff-40e77bd518ed617657533dea53234a6b844143ef5353a5d5f9a234363b1613b8R274

Reply:
resolved: https://github.com/brave/brave-core/pull/7557/files#diff-40e77bd518ed617657533dea53234a6b844143ef5353a5d5f9a234363b1613b8R308